projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-users-endpoint.php

```diff
@@ -248,12 +236,88 @@ function ( $a, $b ) {
 						);
 					}
 
-					$return[ $key ] = $combined_users;
+					$response[ $key ] = $combined_users;
+					break;
+				// Total users from the site. Only available to users with the `list_users` capability (Super admins and admins).
+				case 'total_users':
+					if ( ! empty( $args['total_users'] ) && true === $args['total_users'] && $can_list_users ) {
+						$viewer_count = 0;
+						if ( $include_viewers ) {
+							$viewer_count = $this->get_viewers_count( $viewers, $user_query->get_results(), $blog_id );
+						}
+						$user             = count_users( 'time', $blog_id );
+						$total_users      = isset( $user['total_users'] ) ? (int) $user['total_users'] : 0;
+						$response[ $key ] = $total_users + $viewer_count;
+					}
 					break;
 			}
 		}
 
-		return $return;
+		return $response;
+	}
+
+	/**
+	 * Get the count of private blog users.
+	 *
+	 * @param array $viewers - the viewers.
+	 * @param int   $blog_id - the blog ID.
+	 * @return int
+	 */
+	protected function get_viewers_count( $viewers, $blog_id ) {
+		/*
+		 * We can't get the count of viewers from get_count_private_blog_users because it returns a count for all viewers found,
+		 * even if the viewer has been made a site user. So we want to exclude from the total count users who are viewers.
+		 * Reason is, prior to https://github.a8c.com/Automattic/wpcom/pull/173734, viewers who were promoted to users on a site
+		 * were not removed from drama_blog_access. This function supports the legacy behavior by excluding viewers who are users from the count.
+		 */
+		$duplicate_user_ids = array();
+		$viewers            = get_private_blog_users( $blog_id );
```

Comment:
Just here as an example.

See: https://github.com/Automattic/wp-calypso/pull/100674#issuecomment-2785687150

Reply:
Maybe we could cache all this and set up some hooks to invalidate:


Example, for users:

do_action( 'remove_user_from_blog', $user_id, $blog_id, $reassign );

do_action( ‘add_user_to_blog’, int $user_id, string $role, int $blog_id );


Similar hooks exist for viewers